### ISSUE TITLE
Panic error occurs in v5/pgtype.FlatArray

### ISSUE THREAD

**Issue by jnst** [jackc/pgx#1263]
## Describe the bug

A test case that worked in v4 becomes panic error in v5.
Is this a bug? Or is it a specification?

## To Reproduce

```
--- FAIL: TestScanAll (0.01s)
    --- FAIL: TestScanAll/slice_of_slices_by_ptr (0.00s)
panic: value method github.com/jackc/pgx/v5/pgtype.FlatArray[...].ScanIndexType called using nil *FlatArray[...] pointer [recovered]
```

* error occurs in my forked repo
  * https://github.com/jnst/scany/runs/7532760576?check_suite_focus=true#step:7:254
* test case
  * https://github.com/jnst/scany/blob/master/dbscan/dbscan_test.go#L187-L201
    * `go test -run TestScanAll ./...`
* related commit on pgx v5
  * https://github.com/jackc/pgx/commit/f1a4ae307092a3caedb42ec417ff9be4851779ed

## Version
 - Go: `$ go version` -> `go version go1.18.4 darwin/amd64`
 - PostgreSQL: N/A
 - pgx: `$ grep 'github.com/jackc/pgx/v[0-9]' go.mod` -> `github.com/jackc/pgx/v5 v5.0.0-alpha.5`

**Comment by jackc** [comment]
Fixed in c3258b7f52a3144ecee26228491d7513cd673715.

However, the SQL for that test confuses me. I might be missing something obvious, but it appears that SQL isn't valid. This is what I get when I run it in psql:

```
jack@[local]:5432 pgx_test=# select version();
                                                      version
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 PostgreSQL 14.4 on x86_64-apple-darwin21.5.0, compiled by Apple clang version 13.1.6 (clang-1316.0.21.2.5), 64-bit
(1 row)

Time: 0.601 ms
jack@[local]:5432 pgx_test=# SELECT *
                                FROM (
                                        VALUES (ARRAY('foo val', 'foo val 2')),
                                                (NULL),
                                                (ARRAY('foo val 5', 'foo val 6'))
                                ) AS t (foo);
ERROR:  42601: syntax error at or near "'foo val'"
LINE 3:      VALUES (ARRAY('foo val', 'foo val 2')),
                           ^
LOCATION:  scanner_yyerror, scan.l:1180
Time: 0.258 ms
```

How does that work in the test?

**Comment by jnst** [comment]
I don't understand the source code either, as it is a repository forked for pgx v5 support.

Thanks for the fix!
